src/ring/adapter/jetty9/common.clj

```diff
@@ -1,36 +1,36 @@
 (ns ring.adapter.jetty9.common
-  (:require [clojure.string :as string])
-  (:import [jakarta.servlet.http HttpServletRequest HttpServletResponse]
+  (:require [clojure.string :as string]
+            [ring.core.protocols :as protocols])
+  (:import [org.eclipse.jetty.http HttpHeader HttpField MimeTypes]
+           [org.eclipse.jetty.io.content
+            ContentSourceInputStream ContentSinkOutputStream]
+           [org.eclipse.jetty.server Request Response]
+
            [java.util Locale]))
 
 (defprotocol RequestMapDecoder
   (build-request-map [r]))
 
-(defn set-headers
-  "Update a HttpServletResponse with a map of headers."
-  [^HttpServletResponse response, headers]
-  (doseq [[key val-or-vals] headers]
-    (if (string? val-or-vals)
-      (.setHeader response key val-or-vals)
-      (doseq [val val-or-vals]
-        (.addHeader response key val))))
-  ; Some headers must be set through specific methods
-  (some->> (get headers "Content-Type")
-           (.setContentType response)))
+(defn set-headers!
+  "Update response with a map of headers."
+  [^Response response headers]
+  (let [header-writer (.getHeaders response)]
+    (doseq [[key val-or-vals] headers]
+      (if (string? val-or-vals)
+        (.add header-writer key val-or-vals)
+        (doseq [val val-or-vals]
+          (.add header-writer key val))))))
 
 (defn- header-kv*
-  [^HttpServletRequest req ^String header-name]
-  [(.toLowerCase header-name Locale/ENGLISH)
-   (->> (.getHeaders req header-name)
-        enumeration-seq
-        (string/join ","))])
+  [^HttpField header]
+  [(.getLowerCaseName header) (.getValue header)])
 
 (defn get-headers
   "Creates a name/value map of all the request headers."
-  [^HttpServletRequest request]
-  (->> (.getHeaderNames request)
-       enumeration-seq
-       (into {} (map (partial header-kv* request)))))
+  [^Request request]
+  (let [headers (.getHeaders request)]
+    (into {} (map header-kv* headers))))
```

Comment:
you're using lazyseq here, which may have a 'synchronized' that's incompatible with virtual threads

You nearly wrote a transducer there 

`(into {} (map header-kv*) (.getHeaders request))` should work

Reply:
Nice hint